seacatauth/openidconnect/client/service.py

```diff
@@ -0,0 +1,444 @@
+import datetime
+import logging
+import secrets
+import urllib.parse
+
+import asab.storage.exceptions
+import asab.exceptions
+
+from .. import exceptions
+
+#
+
+L = logging.getLogger(__name__)
+
+#
+
+# TODO: Implement support for remaining metadata
+# https://openid.net/specs/openid-connect-registration-1_0.html#ClientMetadata
+GRANT_TYPES = ["authorization_code", "implicit", "refresh_token"]
+RESPONSE_TYPES = ["code", "id_token", "token"]
+APPLICATION_TYPES = ["web", "native"]
+TOKEN_ENDPOINT_AUTH_METHODS = [
+	"none", "client_secret_basic", "client_secret_post", "client_secret_jwt", "private_key_jwt"]
+CLIENT_METADATA_SCHEMA = {
+	"type": "object",
+	"required": ["redirect_uris"],
+	"additionalProperties": False,
+	"properties": {
+		"redirect_uris": {
+			"type": "array", "description": "Array of Redirection URI values used by the Client."},
+		"client_name": {  # Can have language tags (e.g. "client_name#cs")
+			"type": "string"},
+		#  "contacts": {},
+		"application_type": {
+			"type": "string",
+			"enum": APPLICATION_TYPES},
+		"response_types": {
+			"type": "array",
+			"items": {
+				"type": "string",
+				"enum": RESPONSE_TYPES}},
+		"grant_types": {
+			"type": "array",
+			"items": {
+				"type": "string",
+				"enum": GRANT_TYPES}},
+		# "logo_uri": {},  # Can have language tags
+		"client_uri": {  # Can have language tags
+			"type": "string"},
+		# "policy_uri": {},  # Can have language tags
+		# "tos_uri": {},  # Can have language tags
+		# "jwks_uri": {},
+		# "jwks": {},
+		# "sector_identifier_uri": {},
+		# "subject_type": {},
+		# "id_token_signed_response_alg": {},
+		# "id_token_encrypted_response_alg": {},
+		# "id_token_encrypted_response_enc": {},
+		# "userinfo_signed_response_alg": {},
+		# "userinfo_encrypted_response_alg": {},
+		# "userinfo_encrypted_response_enc": {},
+		# "request_object_signing_alg": {},
+		# "request_object_encryption_alg": {},
+		# "request_object_encryption_enc": {},
+		"token_endpoint_auth_method": {
+			"type": "string",
+			"enum": TOKEN_ENDPOINT_AUTH_METHODS},
+		# "token_endpoint_auth_signing_alg": {},
+		# "default_max_age": {},
+		# "require_auth_time": {},
+		# "default_acr_values": {},
+		# "initiate_login_uri": {},
+		# "request_uris": {},
+		"custom_data": {  # NON-CANONICAL
+			"type": "object", "description": "Additional client data."},
+		"logout_uri": {  # NON-CANONICAL
+			"type": "string", "description": "URI that will be called on session logout."},
+	},
+	# "patternProperties": {
+	#   # Language-specific metadata with RFC 5646 language tags
+	# 	"^client_name#[-a-zA-Z0-9]+$": {"type": "string"},
+	# 	"^logo_uri#[-a-zA-Z0-9]+$": {"type": "string"},
+	# 	"^client_uri#[-a-zA-Z0-9]+$": {"type": "string"},
+	# 	"^policy_uri#[-a-zA-Z0-9]+$": {"type": "string"},
+	# 	"^tos_uri#[-a-zA-Z0-9]+$": {"type": "string"},
+	# }
+}
+
+
+class ClientService(asab.Service):
+	"""
+	Provides API for the registration and management of OpenID Connect clients
```

Comment:
Please specify the name of the standard which this service implements.
Together with a canonical URL to the specifications.